src/OpenTelemetry/Internal/HostingHelper.cs

```diff
@@ -0,0 +1,244 @@
+// <copyright file="HostingHelper.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+using System.Diagnostics;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Internal;
+
+internal static class HostingHelper
+{
+    private static readonly object LockObject = new();
+    private static bool initialized;
+    private static Type? hostedServiceImplementation;
+
+    public static void AddOpenTelemetryHostedServiceIntoServiceCollection(IServiceCollection services)
+    {
+        if (TryAddOpenTelemetryHostedServiceIntoServiceCollection(services, out var reason))
+        {
+            OpenTelemetrySdkEventSource.Log.HostedServiceRegistered();
+        }
+        else
+        {
+            OpenTelemetrySdkEventSource.Log.HostedServiceRegistrationSkipped(reason);
+        }
+    }
+
+    private static bool TryAddOpenTelemetryHostedServiceIntoServiceCollection(IServiceCollection services, out string? reason)
+    {
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+        bool isDynamicCodeSupported = RuntimeFeature.IsDynamicCodeSupported;
+#else
+        // Note: This is for .NET Framework and/or .NET Standard 2.0 targets.
+        bool isDynamicCodeSupported = true;
+#endif
+        if (!isDynamicCodeSupported)
+        {
+            reason = "Dynamic code not supported";
+            return false;
+        }
+
+        var iHostedServiceType = Type.GetType(
+            "Microsoft.Extensions.Hosting.IHostedService, Microsoft.Extensions.Hosting.Abstractions", throwOnError: false);
+
+        if (iHostedServiceType == null)
```

Comment:
How will it behave for AzureFunctions. I believe IHostedService will be found on AzureFunctions but the hostedService infra is not supported.

Reply:
I think Functions team added special support for OTel Hosted Service. I'll try to find that PR, but yes this is something to be tried out. (Independent of this, Otel 1.4 wont work in Functions unless dedicated plan, due to Functions not supporting DS 7.0)

Reply:
So, Otel is going to break for Azure functions with 1.4 upgrade? Shouldn't there be a major version bump if it's breaking the functionality for a group of customers?

Reply:
I do not think so. There is no breaking change in 1.3 -> 1.4. Its just Azure Functions has a strict DS version dependency. As mentioned in the issue in Azure Functions host, the issue is not with OTel. Any app with DS 7.0 will not work.
https://github.com/Azure/azure-functions-host/issues/8938

(IIRC, this issue was there from 1.0 of OTel onwards (https://github.com/Azure/azure-functions-host/issues/7135) and it got fixed, but again broken (with every new .NET release). Hopefully by the time 1.4 is stable, Functions will fix it :)

Either way, lets use https://github.com/open-telemetry/opentelemetry-dotnet/issues/4047 for this conversation, as this is unrelated to this PR.

Reply:
> Independent of this, Otel 1.4 wont work in Functions unless dedicated plan, due to Functions not supporting DS 7.0

For OpenTelemetry .NET customer's point of view it's a breaking change. i.e. Customer using 1.3 upgrades to 1.4, they won't expect any functionality to break or any breaking changes at compile time because this isn't a major version upgrade. So, from OpenTelemetry .NET library's upgrading to DS 7.0 is resulting in a breaking change for it's customer.

Reply:
>  Customer using 1.3 upgrades to 1.4, they won't expect any functionality to break or any breaking changes at compile time

^ There is none.

Reply:
> > Independent of this, Otel 1.4 wont work in Functions unless dedicated plan, due to Functions not supporting DS 7.0
> 
> For OpenTelemetry .NET customer's point of view it's a breaking change. i.e. Customer using 1.3 upgrades to 1.4, they won't expect any functionality to break or any breaking changes at compile time because this isn't a major version upgrade. So, from OpenTelemetry .NET library's upgrading to DS 7.0 is resulting in a breaking change for it's customer.

This I am saying not only for Azure functions but for all other customers

Reply:
There is no breaking change for all customers!

Reply:
If the issue is only with azure functions than it's okay. If it hits any regular .NET flavors than it's a major breaking change.

Reply:
This is a great callout. Shamefully, I was not aware of this in Azure Functions! Playing with the code right now, I think we can make it backwards compatible with what functions is doing.

Reply:
OK I just pushed an update that makes Azure Functions happy locally. Need to verify on an actual build of the packages but I think we're good.

Reply:
@RohitRanjanMS Would you be able to help confirm if this will continue to work in Functions?

Reply:
Thanks @cijothomas, yes, this will work for Functions.